src/Control/Monad/Metrics/Internal.hs

```diff
@@ -25,45 +36,87 @@ import           System.Metrics.Distribution (Distribution)
 import           System.Metrics.Gauge        (Gauge)
 import           System.Metrics.Label        (Label)
 
+type IdenticalKeyTypes m m' = (CounterKey m ~ CounterKey m', GaugeKey m ~ GaugeKey m', DistributionKey m ~ DistributionKey m', LabelKey m ~ LabelKey m')
+type ValidMetricKeys m =
+    ( MetricKey (CounterKey m)
+    , MetricKey (GaugeKey m)
+    , MetricKey (DistributionKey m)
+    , MetricKey (LabelKey m)
+    )
+
+-- | A type can be an instance of 'MonadMetrics' if it can provide a 'Metrics'
+-- somehow. Commonly, this will be implemented as a 'ReaderT' where some
+-- field in the environment is the 'Metrics' data.
+--
+-- * /Since v0.1.0.0/
+class Monad m => MonadMetrics m where
+    type CounterKey      m :: *
+    type GaugeKey        m :: *
+    type DistributionKey m :: *
+    type LabelKey        m :: *
+
+    type CounterKey m      = Text
+    type GaugeKey m        = Text
+    type DistributionKey m = Text
+    type LabelKey m        = Text
+
+    getMetrics :: m (Metrics m)
+
+instance {-# OVERLAPPABLE #-} (MonadMetrics m, MonadTrans t, Monad (t m), IdenticalKeyTypes m (t m)) => MonadMetrics (t m) where
+    getMetrics = liftMetrics <$> lift getMetrics
+
+instance (Monad m, IdenticalKeyTypes m (ReaderT (Metrics m) m)) => MonadMetrics (ReaderT (Metrics m) m) where
+    getMetrics = liftMetrics <$> ask
+
+
+class (Eq key, Hashable key) => MetricKey key where
+    toText :: key -> Text
```

Comment:
I think we can keep this as a class alias, and avoid the Text conversion: eg `class (Eq key, Hashable key) => MetricKey key` and `instance (Eq key, Hashable key) => MetricKey key`.

This is much more in line with what i'm thinking though, thanks :D

Reply:
EKG metric keys are still `Text`, though. At some point we have to convert our typesafe keys into stringly-typed ones for EKG's consumption, no?

Reply:
ah lol, we do! good catch